### PR TITLE
Bring fuzzy matching support into master

### PR DESCRIPTION
**Pull request by hjelmn**
This PR brings the fuzzy matching support developed by @mdosanjh at Sandia into pml/ob1 on master.

The fuzzy matching code is disabled by default and can be enabled on the appropriate platforms by specifying the --with-pml-ob1-matching configure option.

### CODE REVIEW

File: ompi/mca/pml/ob1/custommatch/pml_ob1_custom_match_arrays.h [hunk 1]
```diff
@@ -0,0 +1,589 @@
...
+#include <immintrin.h>
+
+#include "../pml_ob1_recvreq.h"
+#include "../pml_ob1_recvfrag.h"
+
+#define PRQ_SIZE 2
+
+typedef struct custom_match_prq_node
+{
+    int32_t tags[PRQ_SIZE];
```
Comment (bosilca) [inline ompi/mca/pml/ob1/custommatch/pml_ob1_custom_match_arrays.h]: These fields seems to always be accessed together, but laying out the structure this way guarantee 5 cache misses per access which is rather expensive for such a time critical operation.

File: ompi/mca/pml/ob1/custommatch/pml_ob1_custom_match_arrays.h [hunk 2]
```diff
@@ -0,0 +1,589 @@
...
+}
+
+
+// UMQ below.
+
+#define UMQ_SIZE 3
+
+typedef struct custom_match_umq_node
+{
+    int32_t tags[UMQ_SIZE];
```
Comment (bosilca) [inline ompi/mca/pml/ob1/custommatch/pml_ob1_custom_match_arrays.h]: same comment as for the prq struct.

File: ompi/mca/pml/ob1/custommatch/pml_ob1_custom_match_fuzzy512-byte.h [hunk 3]
```diff
@@ -0,0 +1,594 @@
...
+    custom_match_prq_node* elem = list->head;
+    int i;
+    int8_t key = peer ^ tag;
+    __m512i search = _mm512_set1_epi8(key);
+    while(elem)
+    {
+        result = _mm512_cmpeq_epi8_mask(_mm512_and_epi32(elem->keys, elem->mask), _mm512_and_epi32(search, elem->mask));
+        if(result)
+        {
+            for(i = elem->start; i <= elem->end; i++)
```
Comment (bosilca) [inline ompi/mca/pml/ob1/custommatch/pml_ob1_custom_match_fuzzy512-byte.h]: I would think that looping around the set bits in result will be faster as it saves few branches.

File: ompi/mca/pml/ob1/custommatch/pml_ob1_custom_match_fuzzy512-byte.h [hunk 4]
```diff
@@ -0,0 +1,594 @@
...
+            {
+                if((0x1 << i & result) && elem->value[i])
```
Comment (bosilca) [inline ompi/mca/pml/ob1/custommatch/pml_ob1_custom_match_fuzzy512-byte.h]: is 0x1 really promoted to __mmask64 ?

File: ompi/mca/pml/ob1/custommatch/pml_ob1_custom_match_arrays.h (hunk 1 above)
Comment (mdosanjh) [reply ompi/mca/pml/ob1/custommatch/pml_ob1_custom_match_arrays.h]: I agree that restructuring would be pertinent to allow for performant PRQ_SIZE and UMQ_SIZE resizing. The current implementation of 'pml_ob1_custom_match_arrays.h' matching engine structure is explicitly sized to fit each prq/umq node into a single cache line.

**Comment by thananon** [comment]
I have error trying to compile with new matching on Intel Xeon processor with AVX and AVX2 instruction set. GCC 7.1.

Is there something wrong on the configure or am I missing something?

```
/sw/gcc/7.1.0/lib/gcc/x86_64-pc-linux-gnu/7.1.0/include/avx512fintrin.h:3573:1: error: inlining failed in call to always_inline '_mm512_set1_epi32': target specific option mismatch
 _mm512_set1_epi32 (int __A)
 ^~~~~~~~~~~~~~~~~
In file included from custommatch/pml_ob1_custom_match.h:49:0,
                 from pml_ob1_comm.h:38,
                 from pml_ob1.c:51:
custommatch/pml_ob1_custom_match_vectors.h:501:22: note: called from here
         elem->srcs = _mm512_set1_epi32(~0);
                      ^~~~~~~~~~~~~~~~~~~~~
In file included from /sw/gcc/7.1.0/lib/gcc/x86_64-pc-linux-gnu/7.1.0/include/immintrin.h:45:0,
                 from custommatch/pml_ob1_custom_match_vectors.h:17,
                 from custommatch/pml_ob1_custom_match.h:49,
                 from pml_ob1_comm.h:38,
                 from pml_ob1.c:51:
/sw/gcc/7.1.0/lib/gcc/x86_64-pc-linux-gnu/7.1.0/include/avx512fintrin.h:3573:1: error: inlining failed in call to always_inline '_mm512_set1_epi32': target specific option mismatch
 _mm512_set1_epi32 (int __A)
```

**Comment by mdosanjh** [comment]
@thananon The vector code here is writen for AVX-512 and is not currently compatible with AVX or AVX2 (or the limited AVX-512 implementation on Knights Corner).

**Comment by thananon** [comment]
Ah, I see. Thank you.

**Pull request merged**